src_lockfree/mpsc_queue.ml

```diff
@@ -44,10 +44,10 @@ let rec pop t =
   | x :: xs -> t.head <- xs; x
   | Closed -> raise Closed
   | Open ->
-    (* We know the tail is open because we just saw the head was open
-       and we don't run concurrently with [close]. *)
+    (* We know the tail is open because we just saw the head was open *)
+    (* we can run concurrently with close *)
     match Atomic.exchange t.tail Open with
-    | Closed -> assert false
+    | Closed -> pop t
```

Comment:
Hmm... This shouldn't happen when the queue is used correctly.  Note that `close` can no longer be called by producers and the existing tests might be assume that it is safe to call `close` from producers.

Instead of an `assert false` it might be useful to use `failwith` here with a short error message that explains that this shouldn't happen unless `close` was called from a producer.